From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Alan D. Tse" <alandtse@gmail.com>
Date: Sun, 4 Aug 2024 20:42:03 -0700
Subject: [PATCH] 6788: fix: handle namespaces when processing bitfields

Handle cases where namespaces may be provided in a pdb member name. The
code incorrectly assumed that namespaces may never appear if PDB.kind
was a Member.
---
 .../util/bin/format/pdb/DefaultPdbMember.java    | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/DefaultPdbMember.java b/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/DefaultPdbMember.java
index e5de9fa215..1f9528487b 100644
--- a/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/DefaultPdbMember.java
+++ b/Ghidra/Features/PDB/src/main/java/ghidra/app/util/bin/format/pdb/DefaultPdbMember.java
@@ -80,18 +80,18 @@ public class DefaultPdbMember extends PdbMember {
 		if (name == null) {
 			return null;
 		}
+		// name may contain namespace prefix class members
+		int lastColonIndex = name.lastIndexOf("::");
+		if (lastColonIndex > 0) {
+			name = name.substring(lastColonIndex + 2);
+		}
 		if (kind == PdbKind.MEMBER) {
 			// Strip bitfield data if present (see parseBitField method)
 			int bitFieldColonIndex = name.indexOf(':');
 			if (bitFieldColonIndex >= 0) {
-				return name.substring(0, bitFieldColonIndex);
+				name = name.substring(0, bitFieldColonIndex);
 			}
 		}
-		// name may contain namespace prefix for non-Member class members
-		int lastColonIndex = name.lastIndexOf(':');
-		if (lastColonIndex > 0) {
-			name = name.substring(lastColonIndex + 1);
-		}
 		return name;
 	}
 
@@ -122,6 +122,10 @@ public class DefaultPdbMember extends PdbMember {
 		if (name == null || kind != PdbKind.MEMBER) {
 			return;
 		}
+		int namespaceIndex = name.lastIndexOf("::");
+		if (namespaceIndex != -1){
+			name = name.substring(namespaceIndex + 2);
+		}
 		int bitFieldColonIndex = name.indexOf(':');
 		if (bitFieldColonIndex >= 0) {
 
-- 
2.45.1

